### PR TITLE
Parameter: hasDefault, setDefault, and other default functions behaviour fixed

### PR DESCRIPTION
**Pull request by colehurwitz**
Closes: #846 

This changes the behaviour of 'default' in the Parameter class. Basically, if you don't pass in a default value, then Parameter.hasDefault returns False. This will be set to Parameter.hasDefault will return True when the default value is changed (setDefault).

### CODE REVIEW

**Comment by 2xB** [comment]
From a quick look, it seems as if your changes would also work if you did not introduce has_default, as you track self.opts["default"] to be synced with has_default. Is that right?

Btw. for the next time it would probably be a good idea to configure your editor not to remove whitespace, as that would make your Pull Request easier to read.

**Comment by ixjlyons** [comment]
I agree -- I think we *either* want `has_default` with `self.opts['default']` falling back to `None` *or* `'default' in self.opts` logic if a `'default'` key isn't added implicitly. I tend to have a slight preference for avoiding adding another attribute.

As a note for @2xB: there's an option in the github "files changed" tab to hide whitespace changes, though I tend to agree we should keep linting-type changes minimal outside dedicated PRs for things like that.

**Comment by colehurwitz** [comment]
Thanks for the feedback @2xB and @ixjlyons . For my intended usage, I actually needed the has_default boolean. I was assuming that a user could pass in None as a default value so in that case self.opts['default']=None but has_default=False. Is that an incorrect assumption?

Sorry about the white space! I will figure that out.

**Comment by ixjlyons** [comment]
Hmm, I considered the case where the user actually wants the default to be `None`, so explicitly passing `default=None` will work (`self.opts['default']` gives `None`, `hasDefault` returns `True`). I'm confused why you'd pass `default=None` but still want `hasDefault` to return `False` though -- can't you just not pass the `default` keyword argument if you don't want a default? I may be missing something here; I haven't really used the parametertree stuff much.

**Comment by colehurwitz** [comment]
Ah, sorry, I screwed up that explanation (it has been a while since I made the change). I meant to say that I wanted to differentiate between when default was set to None and when default is set to None because not default value was passed in. 

User passes in None for default --> has_default=True and default=None
User does not pass in default value --> has_default=False and default=None

So if you just tracked whether or not default=None, then you can not differentiate between these cases. This is why I introduced the has_default boolean.

**Comment by ixjlyons** [comment]
Right, so the approach I was suggesting is to *not* add a `'default'` key at all to `self.opts` when no default has explicitly been set. Then `hasattr(self.opts, 'default')` accurately reflects whether or not the user explicitly passed a default value (could be `None` or whatever) and `defaultValue` returns `None` as a fallback if there isn't a default value. So:

- user passes `default=None`:
    - `hasattr(self.opts, 'default')` -> `True`, so `hasDefault` -> `True`
    - `defaultValue` -> `self.opts['default']` which is `None`
- user does not pass `default` keyword arg:
    - `hasattr(self.opts, 'default')` -> `False`, so `hasDefault` -> `False`
    - `defaultValue` -> `None` as fallback

File: pyqtgraph/parametertree/Parameter.py [hunk 1]
```diff
@@ -408,14 +417,17 @@ def setToDefault(self):
 
     def hasDefault(self):
         """Returns True if this parameter has a default value."""
-        return 'default' in self.opts
-        
+        return self.has_default
```
Comment (2xB) [inline pyqtgraph/parametertree/Parameter.py]: How about just returning ˋself.defaultValue() is not Noneˋ?

**Comment by 2xB** [comment]
I would not differenciate between no default and a default of ˋNoneˋ, as the latter is the most common to explicitly describe the former.

File: pyqtgraph/parametertree/Parameter.py (hunk 1 above)
Comment (2xB) [reply pyqtgraph/parametertree/Parameter.py]: (Instead of using has_default etc.?)

**Pull request closed without merging**